contracts/exchange/test/transactions_unit_tests.ts

```diff
@@ -0,0 +1,694 @@
+        });
+
+        // FIXME - This should be unskipped when the contracts have been updated to fix this problem.
+        it.skip('should revert if reentrancy occurs in the middle of an executeTransaction call and msg.sender == signer and then msg.sender != sender', async () => {
+            const validSignature = randomSignature();
+
+            const currentTimestamp = await getLatestBlockTimestampAsync();
+            const transaction1 = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
+                data: getExecutableCallData(true, constants.NULL_BYTES, constants.NULL_BYTES), // This should never get called
+                signerAddress: accounts[0],
+                domain,
+            };
+            const transactionHash1 = transactionHashUtils.getTransactionHashHex(transaction1);
+
+            const callData = getExecutableCallData(
+                true,
+                getExecuteTransactionCallData(transaction1, validSignature),
+                '0xdeadbeef',
+            );
+
+            const transaction2 = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
+                data: callData,
+                signerAddress: accounts[1],
+                domain,
+            };
+            const transactionHash2 = transactionHashUtils.getTransactionHashHex(transaction2);
+            const abiEncoder = AbiEncoder.createMethod('TransactionError', ['uint8', 'bytes32']);
+            const errorData = abiEncoder.encode([
+                ExchangeRevertErrors.TransactionErrorCode.NoReentrancy,
+                transactionHash1,
+            ]);
+
+            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(transactionHash2, errorData);
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction2, validSignature, {
+                from: accounts[1], // Different then the signing addresses
+            });
+            return expect(tx).to.revertWith(expectedError);
+        });
+
+        it('should revert if the transaction has been executed previously', async () => {
+            const validSignature = randomSignature();
+            const currentTimestamp = await getLatestBlockTimestampAsync();
+            const transaction = {
+                ...EMPTY_ZERO_EX_TRANSACTION,
+                expirationTimeSeconds: new BigNumber(currentTimestamp).plus(10),
+                data: getExecutableCallData(true, constants.NULL_BYTES, constants.NULL_BYTES),
+                domain,
+            };
+            const transactionHash = transactionHashUtils.getTransactionHashHex(transaction);
+
+            // Use the transaction in execute transaction.
+            await expect(
+                transactionsContract.executeTransaction.sendTransactionAsync(transaction, validSignature),
```

Comment:
This technically doesn't wait for the tx to be mined (even though it's probably all the same on ganache).
```suggestion
                transactionsContract.executeTransaction.awaitTransactionSuccessAsync(transaction, validSignature),
```
There's a couple other spots where this is done, too.